rules/DeadCode/Rector/FunctionLike/NarrowTooWideReturnTypeRector.php

```diff
@@ -0,0 +1,250 @@
+
+        if (! $returnType instanceof UnionType) {
+            return null;
+        }
+
+        $actualReturnTypes = $this->collectActualReturnTypes($node, $returnStatements, $isAlwaysTerminating);
+        $newReturnType = $this->narrowUnionReturnType($returnType, $actualReturnTypes);
+
+        if ($newReturnType === null) {
+            return null;
+        }
+
+        $node->returnType = $newReturnType;
+
+        return $node;
+    }
+
+    private function shouldSkipNode(ClassMethod|Function_|Closure|ArrowFunction $node, Scope $scope): bool
+    {
+        if ($this->hasYield($node)) {
+            return true;
+        }
+
+        if (! $node instanceof ClassMethod) {
+            return false;
+        }
+
+        if ($node->isPrivate() || $node->isFinal()) {
+            return false;
+        }
+
+        if ($node->isAbstract()) {
+            return true;
+        }
+
+        $classReflection = $this->reflectionResolver->resolveClassReflection($node);
+
+        if (! $classReflection instanceof ClassReflection) {
+            return true;
+        }
+
+        return ! $classReflection->isFinalByKeyword();
+    }
+
+    /**
+     * @param Return_[] $returnStatements
+     * @return Type[]
+     */
+    private function collectActualReturnTypes(
+        ClassMethod|Function_|Closure|ArrowFunction $node,
+        array $returnStatements,
+        bool $isAlwaysTerminating,
+    ): array {
+        if ($node instanceof ArrowFunction) {
+            return [$this->nodeTypeResolver->getNativeType($node->expr)];
+        }
+
+        $returnTypes = [];
+        foreach ($returnStatements as $returnStatement) {
+            if ($returnStatement->expr === null) {
+                $returnTypes[] = new NullType();
+                continue;
+            }
+
+            $returnTypes[] = $this->nodeTypeResolver->getNativeType($returnStatement->expr);
+        }
+
+        if (! $isAlwaysTerminating) {
+            $returnTypes[] = new NullType();
+        }
+
+        return $returnTypes;
+    }
+
+    /**
+     * @param Type[] $actualReturnTypes
+     */
+    private function narrowUnionReturnType(
+        UnionType $unionType,
+        array $actualReturnTypes
+    ): ComplexType|Identifier|Name|null {
+        $types = $unionType->types;
+        $usedTypes = [];
+
+        foreach ($types as $type) {
+            $declaredType = $type instanceof Expr
+                ? $this->nodeTypeResolver->getNativeType($type)
+                : $this->getType($type);
+
+            if ($declaredType instanceof MixedType) {
+                // Mixed type covers all other types, so we should only keep mixed
+                return new Identifier('mixed');
+            }
+
+            foreach ($actualReturnTypes as $actualType) {
+                if (! $declaredType->isSuperTypeOf($actualType)->no()) {
+                    $usedTypes[] = $declaredType;
+                    break;
+                }
+            }
+        }
+
```

Comment:
$usedTypes may need to have unique check:

```suggestion
            $usedTypes = array_unique($usedTypes, SORT_REGULAR);
```

that ensure repetive types not recount:

```php
function foo(): int|null
{
    if (rand(0, 1)) {
         return 1;
   } 

    if (rand(0, 1)) {
         return 2;
   } 

    if (rand(0, 1)) {
         return 3;
   } 

  return null;
}
```

Reply:
It seems that I did not need this (the test passed without reapplying the rule with no changes) but I can add to be on the safe side